src/range.cr

```diff
@@ -392,12 +392,78 @@ struct Range(B, E)
       raise ArgumentError.new("Can't sample an open range")
     end
 
-    return super unless n == 1
+    if n < 0
+      raise ArgumentError.new "Can't sample negative number of elements"
+    end
 
-    if empty?
+    case n
+    when 0
       [] of B
-    else
+    when 1
       [sample(random)]
+    else
+      # For a range of integers we can do much better
+      {% if B < Int && E < Int %}
+        min = self.begin
+        max = self.end
+
+        if (exclusive? && max <= min) || (!exclusive? && max < min)
+          raise ArgumentError.new "Invalid range for rand: #{self}"
+        end
+
+        max -= 1 if self.exclusive?
+
+        available = max - min + 1
+        possible = Math.min(n, available)
+
+        # If we must return all values in the range...
+        if possible == available
+          result = Array(B).new(possible)
+          each { |value| result << value }
+          result.shuffle!
+          result
+        elsif n <= 16
+          # For a small requested amount doing a linear lookup is faster
+          result = Array(B).new(possible)
+          add_n_samples(result, possible, random)
+          result
+        else
+          # Otherwise using a Set is faster
+          result = Set(B).new(possible)
+          add_n_samples(result, possible, random)
+          result.to_a
+        end
+      {% elsif B < Float && E < Float %}
+        min = self.begin
+        max = self.end
+
+        if (exclusive? && max <= min) || (!exclusive? && max < min)
+          raise ArgumentError.new "Invalid range for rand: #{self}"
+        end
+
+        if min == max
+          [min]
+        elsif n <= 16
```

Comment:
Maybe use `Array::SMALL_ARRAY_SIZE`?

Reply:
It's actually `Hash::MAX_INDICES_SIZE_LINEAR_SCAN / 2` but it's a private constant.

Regardless, the heuristic of using a small number (whatever that number is) is independent of the actual values using in other types.

Reply:
`Array::SMALL_ARRAY_SIZE` is used for deciding between linear scan and hash lookup, so that's looks pretty much like the same heuristic use case.

Reply:
Okay... but it's a private constant.

Reply:
Sure we can change it to be public + nodoc.

Reply:
From https://go-proverbs.github.io :

> A little copying is better than a little dependency

Reply:
And here it's just a number literal.

Reply:
Definitely agree to that. But the constant gives some context. Now it's just a magic number without any explanation.

Reply:
What do you mean? There's a comment right below that condition.

Reply:
Yeah, I mean without explanation of the specific value.

It's just the same number we use for equivalent situations. So IMO using a common constant would make sense. But I'm not going to press this any further.

Reply:
This constant will never, ever change. So we can continue copying over. The problem would be if we later had to update this number everyone. But this number has been proven to be the optimal one.